File: src/services/notificationService.js
/**
 * notificationService.js
 * --------------------------------------------------------------------------
 * Bildirishnoma servisi. Ikki vazifa:
 *
 *   1) Davriy ravishda toza xonalarni tekshiradi. Agar oxirgi tozalanish
 *      vaqtidan beri "cleaningThresholdHours" (default: 12 soat) o'tgan
 *      bo'lsa, tozalovchi xodimga bildirishnoma yuboradi va xonani
 *      'iflos' (dirty) deb belgilash uchun 'room.cleaning_required'
 *      hodisasini brokerga nashr etadi.
 *
 *   2) Brokerdagi barcha 'notification.created' hodisalarni tinglaydi va
 *      ularni store ga yozadi (panel va telefon ovozli xabarnomasi uchun).
 *
 * Bu servis WebSocket orqali ham xabar yuboradi (wsServer.js qabul qiladi).
 * --------------------------------------------------------------------------
 */

'use strict';

const crypto = require('crypto');
const broker = require('../broker/messageBroker');
const store = require('../data/store');
const logger = require('../utils/logger');

const ONE_HOUR = 1000 * 60 * 60;

// Demo maqsadlarida tezroq tekshirish (1 daqiqada bir marta).
// Real ishlab chiqarishda har 5-10 daqiqada bir tekshirish yetarli.
const CHECK_INTERVAL_MS = 60 * 1000;

function genId() {
  return `notif_${Date.now()}_${crypto.randomBytes(3).toString('hex')}`;
}

class NotificationService {
  constructor() {
    this.name = 'notification';
    this._timer = null;
    this._subscribe();
    this.startPeriodicCheck();
    logger.info('[NOTIFICATION] Servis ishga tushdi');
  }

  _subscribe() {
    broker.subscribe('notification.created', (event) => {
      const settings = store.getSettings();
      const notif = {
        id: genId(),
        type: event.payload.type || 'info',
        severity: event.payload.severity || 'info',
        message: event.payload.message,
        roomNumber: event.payload.roomNumber ?? null,
        orderId: event.payload.orderId ?? null,
        requestId: event.payload.requestId ?? null,
        createdAt: Date.now(),
        read: false,
        soundEnabled: settings.notificationSound,
      };
      store.addNotification(notif);
    });
  }

  /** Davriy tekshiruvni boshlash */
  startPeriodicCheck() {
    if (this._timer) clearInterval(this._timer);
    this._timer = setInterval(() => this.checkCleaningSchedule(), CHECK_INTERVAL_MS);
    // Birinchi tekshiruvni darhol ishga tushiramiz
    setTimeout(() => this.checkCleaningSchedule(), 5000);
  }

  /** Asosiy logika: har 12 soatda tozalash kerakligi xabarnomasi */
  checkCleaningSchedule() {
    const settings = store.getSettings();
    const thresholdMs = (settings.cleaningThresholdHours || 12) * ONE_HOUR;
    const now = Date.now();
    const rooms = store.getRooms();
    let triggered = 0;

    for (const room of rooms) {
      // Hozir tozalanmayotgan, navbatda turmagan va texnik xizmatda bo'lmagan xonalar
      if (room.status === 'cleaning' || room.status === 'maintenance' || room.status === 'cleaning_required' || room.status === 'inspection') {
        continue;
      }
      // Oxirgi tozalanish vaqtidan beri o'tgan vaqt
      const elapsed = now - (room.lastCleanedAt || 0);
      if (elapsed < thresholdMs) continue;

      // Bildirishnoma takrorlanmasligi uchun
      if (room.cleaningReminderSentAt && (now - room.cleaningReminderSentAt) < thresholdMs) {
        continue;
      }
      store.updateRoom(room.number, { cleaningReminderSentAt: now });

      // BAND xonalar uchun: faqat eslatma yuboramiz — mehmon ichida bo'lganda
      // statusni o'zgartirib bo'lmaydi. Tozalovchi mehmon bilan kelishadi.
      if (room.status === 'occupied') {
        broker.publish('notification.created', {
          type: 'occupied_room_cleaning_due',
          severity: 'warning',
          targetRole: 'housekeeping',
          message: `🛎 ${room.number}-xona (BAND) ${Math.round(elapsed / ONE_HOUR)} soatdan beri tozalanmagan. Mehmon bilan kelishib tozalang.`,
          roomNumber: room.number,
        });
        triggered++;
        continue;
      }

      // BO'SH xonalar uchun: yangi mehmon kelishidan oldin tozalash kerak
      // (statusni cleaning_required ga olib o'tamiz)
      broker.publish('room.cleaning_required', {
        roomNumber: room.number,
        elapsedHours: Math.round(elapsed / ONE_HOUR),
        thresholdHours: settings.cleaningThresholdHours,
      });

      broker.publish('notification.created', {
        type: 'cleaning_reminder',
        severity: 'warning',
        targetRole: 'housekeeping',
        message: `🧹 ${room.number}-xona ${Math.round(elapsed / ONE_HOUR)} soatdan beri tozalanmagan. Yangi mehmonlardan oldin tozalash kerak.`,
        roomNumber: room.number,
      });
      triggered++;
    }

    if (triggered > 0) {
      logger.info(`[NOTIFICATION] Tozalash bildirishnomasi yuborildi: ${triggered} xona`);
    }
  }

  stop() {
    if (this._timer) clearInterval(this._timer);
  }
}

module.exports = new NotificationService();

File: backend/app/services/notification_service.py
import time
import secrets
import asyncio
from app.broker.message_broker import broker
from app.data.store import store
from app.utils.logger import logger

ONE_HOUR = 1000 * 60 * 60
CHECK_INTERVAL_MS = 60 * 1000

def gen_id():
    return f"notif_{int(time.time() * 1000)}_{secrets.token_hex(3)}"

class NotificationService:
    def __init__(self):
        self.name = 'notification'
        self._task = None
        self._subscribe()
        logger.info('[NOTIFICATION] Servis ishga tushdi')

    def _subscribe(self):
        def on_notification_created(event):
            settings = store.get_settings()
            payload = event.get('payload', {})
            notif = {
                "id": gen_id(),
                "type": payload.get('type', 'info'),
                "severity": payload.get('severity', 'info'),
                "message": payload.get('message'),
                "roomNumber": payload.get('roomNumber'),
                "orderId": payload.get('orderId'),
                "requestId": payload.get('requestId'),
                "createdAt": int(time.time() * 1000),
                "read": False,
                "soundEnabled": settings.get('notificationSound', True),
            }
            store.add_notification(notif)
            
        broker.subscribe('notification.created', on_notification_created)

    async def _loop(self):
        await asyncio.sleep(5)
        self.check_cleaning_schedule()
        while True:
            await asyncio.sleep(CHECK_INTERVAL_MS / 1000)
            self.check_cleaning_schedule()

    def start_periodic_check(self):
        if self._task:
            self._task.cancel()
        loop = asyncio.get_running_loop()
        self._task = loop.create_task(self._loop())

    def check_cleaning_schedule(self):
        settings = store.get_settings()
        threshold_hours = settings.get('cleaningThresholdHours', 12)
        threshold_ms = threshold_hours * ONE_HOUR
        now = int(time.time() * 1000)
        rooms = store.get_rooms()
        triggered = 0

        for room in rooms:
            if room.get('status') in ('cleaning', 'maintenance', 'cleaning_required', 'inspection'):
                continue
                
            elapsed = now - room.get('lastCleanedAt', 0)
            if elapsed < threshold_ms:
                continue
                
            last_reminder = room.get('cleaningReminderSentAt', 0)
            if (now - last_reminder) < threshold_ms:
                continue
                
            store.update_room(room['number'], {"cleaningReminderSentAt": now})

            if room.get('status') == 'occupied':
                broker.publish('notification.created', {
                    "type": 'occupied_room_cleaning_due',
                    "severity": 'warning',
                    "targetRole": 'housekeeping',
                    "message": f"🛎 {room['number']}-xona (BAND) {round(elapsed / ONE_HOUR)} soatdan beri tozalanmagan. Mehmon bilan kelishib tozalang.",
                    "roomNumber": room['number'],
                })
                triggered += 1
                continue

            broker.publish('room.cleaning_required', {
                "roomNumber": room['number'],
                "elapsedHours": round(elapsed / ONE_HOUR),
                "thresholdHours": threshold_hours,
            })

            broker.publish('notification.created', {
                "type": 'cleaning_reminder',
                "severity": 'warning',
                "targetRole": 'housekeeping',
                "message": f"🧹 {room['number']}-xona {round(elapsed / ONE_HOUR)} soatdan beri tozalanmagan. Yangi mehmonlardan oldin tozalash kerak.",
                "roomNumber": room['number'],
            })
            triggered += 1

        if triggered > 0:
            logger.info(f"[NOTIFICATION] Tozalash bildirishnomasi yuborildi: {triggered} xona")

    def stop(self):
        if self._task:
            self._task.cancel()
            self._task = None

notification_service = NotificationService()
